### DISCUSSION TITLE
Asynchronous validation in use-form hook

### DISCUSSION THREAD

**Discussion by ChapterSevenSeeds** [mantinedev/mantine#3614] [Feature requests]
I think that an option to perform asynchronous form validation would be great. Consider the following Electron code.

```ts
export default function App() {
    const theme = useMantineTheme();

    const form = useForm<{ file: string }>({
        validate: zodResolver(z.object({
            file: z.string().refine(async (path: string) => await window.api.fileExists(path), "File doesn't exist.")
        })),
        initialValues: {
            file: ""
        },
        validateInputOnChange: true
    });

    return (
        <form onSubmit={form.onSubmit(values => {
            console.log(values);
        })}
        >
            <TextInput
                label="Default Connect Path"
                {...form.getInputProps("file")}
            />
            <Button variant='default' type="submit">Save</Button>
        </form >
    );
}
```

Typing a value in the TextInput crashes the window with a Zod error saying, "Async refinement encountered during synchronous parse operation. Use .parseAsync instead." 

Is this a reasonable feature to request?

**Comment by rtivital** [comment]
Mantine form does not current support async validation. This feature is in the backlog.

**Reply by ryuhangyeong** [reply]
Is this feature still not supported?

**Reply by gius** [reply]
Do you have any recommendations on how to properly implement async validation? 
Typical problems:
 - blocking submit while validation is in process
 - reacting to validation result (manually setting error or even manipulating form values) with the actual `form` instance (instead of the one caught in a closure when the handler was originally called)

**Reply by lukepighetti** [reply]
Checking in on this one. Definite need for us. New to the ecosystem though so if anyone has a recommendation for a package or pattern to use to backfill this functionality I'm all ears

**Reply by gius** [reply]
I would try https://tanstack.com/form/latest

**Comment by ivosabev** [comment]
Any progress on this?

**Comment by ryuhangyeong** [comment]
It looks like there have been improvements to form in 7.8. Async validation is also a very common use case, so I really hope Mantine supports it.

**Comment by Gus007-BR** [comment]
This is a huge bummer... I was so happy with the use-form hook... I hope this gets addressed soon. In the meantime, please let us know if you happen to know of any good workarounds.

**Reply by ryuhangyeong** [reply]
https://mantine.dev/changelog/7-9-0/#use-field-hook

**Reply by ivosabev** [reply]
Why was it not implemented on the form level as well?

**Comment by martinestiga** [comment]
Any news about this?? This functionality is really necessary.

**Reply by ryuhangyeong** [reply]
`use-field`

https://mantine.dev/form/use-field/

**Comment by dmgauthier** [comment]
Please consider this basic functionnality at the form level. 
A dev in our team had to do a frankeinstein between Mantine and React Hook Form to implement async validation. :zombie:

**Reply by rtivital** [reply]
How do you perform validation? Is it entirely on the backend or only some fields are validated on backend? Why do you not validate inside `onSubmit` handler? What features/API do you expect from async validation?

**Reply by ryuhangyeong** [reply]
@rtivital 

maybe [here](https://element.eleme.io/#/en-US/component/form#custom-validation-rules)

```js
// @example
const form = useForm({
  mode: 'uncontrolled',
  initialValues: {
    email: '',
    termsOfService: false,
  },

  validate: {
    // Enable asynchronous processing in the validate function
    email: async(value) => {
      // @example
      await mockApi();
      
      return null
    }
  },
});
```

**Reply by ryuhangyeong** [reply]
```js
// @example
const rules = {
  number: [
    { required: true, message: 'error', trigger: 'blur' },
    {
      validator: async (_, value, callback) => {
        const data = await mockApi();
        return callback(data);
      },
      trigger: 'blur',
    },
  ],
}
```

https://github.com/yiminghe/async-validator

**Reply by dmgauthier** [reply]
TLDR: 
> What features/API do you expect from async validation?

[zod superRefine](https://zod.dev/?id=superrefine&ref=fun4code)

Long story short,
Even though we DO have a backend that validates. (We wired up the mantine notification system on a custom client handling errors to show a toast when we end up with a throw there)
We try to validate before sending stuff to the API with zod. And this is precisely in the [zod superRefine](https://zod.dev/?id=superrefine&ref=fun4code). E.G. making a call with filter to check if a resource name is duplicated. 
Having a mixed up implementation between async stuff in onBlur listener+custom validation and our stronger shema base validation is not better than having a RHF frankeinstein IMHO

**Reply by martinestiga** [reply]
In my case, like @dmgauthier , it's to validate if the email of the user that is being created already exists or not.

**Comment by gary-menzel** [comment]
I am plus one hundred on this.  async validation is important any time you need to use a promise - for whatever reason.  It shouldn't need to be justified against other peoples view of HOW validation should be done.

**Comment by davisilvarafacho** [comment]
I also think that asynchronous validation is not only important, but essential! It doesn't make much sense to have a validation hook if it doesn't support a promise, since there are very few cases where nothing depends on an external query.

I believe it's high time that a robust framework like Mantine implements asynchronous validation in its main form object.

**Comment by gary-menzel** [comment]
I am returning to this after having found a solution that works for me...  You can override the whole `form.validate` function wih an async function but you haev to set the errors one by one (with `form.setFieldError`). The process of setting a field error is actually an async-like process as it uses events.  My use case is more complex than most because I have expressions that execute async - so it is is tricky to provide a simple working example.  But here is basically what my setup looks like:
```javascript
const form = useForm({
    validate: (values) => {
      customValidator(form)(values)
    }, // etc...
 })
```
`customValidator` is a function that generates the actual validator function and provides the `form` as scope.  The generated function can be `async`.  The `form` is required so you can do `form.setFieldError` (and, in my case, I can also run my validation rules against any parts of the form (as I said - the model is complex - but this may give others ideas on what they can do).  You can still use or wrap the Mantine validator functions.

**Reply by CristianCrane** [reply]
thanks for sharing, i appreciate your insight

**Reply by KoljaFrahm** [reply]
I don't totally understand how you did it, do you have an example?
`validate` expects a function like  `((values: Values) => FormErrors)`, I don't see how this can be async as FormErrors is not a promise.

**Reply by gary-menzel** [reply]
(sorry I took a while to reply - been busy) - I may have neglected to explain that I only used the validate function as a hook so I could trap each call to it.  I then set the errors on the form directly (by passing in the form to the custom validation function along with the values to validate.  So I return nothing from the `validate` function.  But, I see this as a "hack" and why they should support `async` validation functions.

**Comment by KoljaFrahm** [comment]
I have to agree here that this feature is necessary!

**Comment by franzos** [comment]
I recently migrated some of my apps from client- to server-side validation to reduce code duplication and because the API anyway has to make sure form submissions are correct; I've found it's really easy to do this, without async hook. Here's a basic example:

```js
// Setup
const form = useForm({
    mode: 'controlled',
    initialValues: initialValues as Record<string, unknown>,
});

// Submit handler
const submitForm = async (data: typeof form.values) => {
    try {
      // Callback to the parent component or API call
      await submitFormCb(data);
    } catch (e) {
      // Handle form related errors
      if (hasFieldValidationError(e)) {
        form.setErrors(axiosFieldValidationErrorToFormErrors(e));
      }
    }
};
```

**Reply by dmgauthier** [reply]
Of course, and thanks for sharing this.

But the issue is when using a global shema validation like Zod.
